### DISCUSSION TITLE
Generating changelog without releasing

### DISCUSSION THREAD

**Discussion by someth2say** [jreleaser/jreleaser#367] [Q&A] [answered]
I'm (finally!) starting my journey with JReleaser. It is a wonderful project, @aalmiray !

I am a bit puzzled by a behavior I do not fully understand.
I'm trying to generate the changelog for one of my side projects (https://github.com/someth2say/xmlformat), but still not prepared to release it. 
I started tacking the `changelog` command, but I promptly faced that the command requires a GitHub token to just generate the file:
```
➜  xmlformat git:(releaser) jreleaser init --format yml  
[INFO]  Writing file /Users/jordisola/Projects/someth2say/xmlformat/jreleaser.yml
[INFO]  JReleaser initialized at /Users/jordisola/Projects/someth2say/xmlformat
➜  xmlformat git:(releaser) ✗ jreleaser changelog          
[INFO]  JReleaser 0.6.0
[INFO]  Configuring with jreleaser.yml
[INFO]    - basedir set to /Users/jordisola/Projects/someth2say/xmlformat
[INFO]  Reading configuration
[INFO]  Loading variables from /Users/jordisola/.jreleaser/config.properties
[WARN]  Variables source /Users/jordisola/.jreleaser/config.properties does not exist
[INFO]  Validating configuration
[ERROR] == JReleaser ==
[ERROR] github.token must not be blank. Configure a value using the CLI yaml DSL, or define a JRELEASER_GITHUB_TOKEN environment variable, or define a key/value pair in /Users/jordisola/.jreleaser/config.properties
JReleaser has not been properly configured.
➜  xmlformat git:(releaser) ✗ JRELEASER_GITHUB_TOKEN=empty jreleaser changelog
[INFO]  JReleaser 0.6.0
[INFO]  Configuring with jreleaser.yml
[INFO]    - basedir set to /Users/jordisola/Projects/someth2say/xmlformat
[INFO]  Reading configuration
[INFO]  Loading variables from /Users/jordisola/.jreleaser/config.properties
[WARN]  Variables source /Users/jordisola/.jreleaser/config.properties does not exist
[INFO]  Validating configuration
[INFO]  Project version set to 1.0.0-SNAPSHOT
[INFO]  Release is snapshot
[INFO]  Timestamp is 2021-09-14T22:11:25.516325+02:00
[INFO]  HEAD is at 7c8298e
[INFO]  dryrun set to false
[INFO]  Generating changelog: out/jreleaser/release/CHANGELOG.md
[INFO]  Writing output properties to out/jreleaser/output.properties
[INFO]  JReleaser succeeded after 0.046 s
```
As seen here, simply using a fake token does the trick to generate the file.
Is this somehow expected or desired?

**Comment by aalmiray** [comment, marked as answer]
Hi Jordi, yes this is expected for the time being. JReleaser validates the configuration before invoking a command. In the case of generating a changelog a non-empty git token is required, even if no remote operations (such as resolving username links) are not requested. 

JReleaser could be a bit smarter and avoid requiring a token when no remote operation will be invoked. 

Passing a fake git token as you did is a viable workaround.

**Reply by helpermethod** [reply]
Perfect, just stumbled across the same problem 😄.

**Comment by sualeh** [comment]
@someth2say - please could you share your "jreleaser.xml" file? I do not want to create a tag or release, just a changelog, but I get 

```
JReleaser has not been properly configured.
No release provider has been configured 
```

**Reply by aalmiray** [reply]
The simplest configuration you may  put in place is https://jreleaser.org/guide/latest/examples/simple-release.html

Your project sources have to be inside a git repository with a remote whose name defaults to "origin".

https://jreleaser.org/guide/early-access/checklist.html

**Reply by someth2say** [reply]
I do not have it handy, but it's pretty easy based on @aalmiray's sample file:
```yaml
project:
  name: app
  version: 1.0.0
release:
  github:
    token: __DO_NOT_SET_HERE__
```
Any value in the `release.github.token` field will do.
Then, just running `jreleaser changelog` does the trick.

**Reply by sualeh** [reply]
@aalmiray - why don't the following behave in the same way?

```yaml
release:
  github:
    token: __DO_NOT_SET_HERE__
```

and 

```yaml
release:
  github:
    enabled: false
```

**Reply by sualeh** [reply]
So, instead of the fake token, I ended up doing this instead, and it works. (I would have hoped that the enabled flag would work, but it didn't.

```yaml
release:
  github:
    skipRelease: true
    skipTag: true
```

Of course, I had to set the "JRELEASER_GITHUB_TOKEN" environmental variable, but since I am using GitHub Actions, it was easy enough.
 
```yaml
    - id: jreleaser
      name: Run JReleaser
      uses: jreleaser/release-action@v1
      env:
        JRELEASER_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

**Reply by aalmiray** [reply]
Just had a look at the code and the `enabled` flag is set but not validated. This is likely a missing check. 

Also not having a releaser enabled would cause other problems down the road but I can see the use case for generating the changelog. Mind that even this case would require a releaser to resolve usernames if specified by the formatting options.

**Reply by sualeh** [reply]
@aalmiray - do you think there is a valid use case for not having any releasers? For example, I may want to manage my GitHub release myself, but make announcements using JReleaser.

**Reply by aalmiray** [reply]
If there are no releasers then packages will likely fail unless there's at least one uploader configured. 

A releaser does not necessarily have to create a release, it can also update an existing release. 

However if you only need JReleaser to post announcements without a releaser then that could be arranged. Such configuration would deactivate changelog generation though as right now it depends on a releaser being configured.

**Reply by sualeh** [reply]
@aalmiray  - please take a look at https://github.com/schemacrawler/SchemaCrawler-Installers/blob/master/jreleaser.yml

I do not do changelogs, or creating or updating releases, no packaging - just announcements. It is working. I think this may become a use case for some people.

**Reply by sualeh** [reply]
(I understand that `releaseNotesUrl` needs to be computed from GitHub.)

**Reply by aalmiray** [reply]
FWIW a releaser is core to the behavior, even if not active as you noted some URLs would still be used by other parts of the tool.

**Reply by sualeh** [reply]
So then, if all I want to do is make announcements, then what I have in my config if fine, since it satisfies my use case. Also, I am not using any flaky `token` hacks; and I cannot set `enabled: false`, since then I will not get `releaseNotesUrl`.  

Do you have any additional thoughts?

**Reply by aalmiray** [reply]
Correct. Once `enabled: false` is honored you'd still miss something. Setting the configuration as you have it right now is OK.
